### PR TITLE
Il-to-Il pass: Eliminate the ! operator equationally

### PR DESCRIPTION
**Pull request by nomeata**
This pass can go before the sideconditions pass #9, to do something more clever with option projection. Here is an example:

```diff
   ;; 3-typing.watsup:238.1-240.23
-  rule extend {C : context, n : n, nt : numtype}:
+  rule extend {C : context, n : n, nt : numtype, o0 : nat}:
     `%|-%:%`(C, EXTEND_instr(nt, n), `%->%`([(nt <: valtype)], [(nt <: valtype)]))
-    -- if ($size(nt <: valtype) =/= ?())
-    -- if (n <= !($size(nt <: valtype)))
+    -- if ($size(nt <: valtype) = ?(o0))
+    -- if (n <= o0)
```

Before it would introduce `!= null` constraints and leave `!(e)` in place, which is correct, but typically hard to work with in a formalization. With this pass before, it introduces a fresh variable `o0` and asserts that the partial expression is `= ?o0`.

This is a separate pass because maybe only some backends will prefer this (and because it’s slightly complicated).

### CODE REVIEW

**Comment by nomeata** [comment]
I did some mildly nasty things to `Validation.mli` to be able to infer the types of expressions. This probably needs some cleanup there first (immutable `validate_bind` for example), or change validation to add type annotations altogether, or so.

**Comment by nomeata** [comment]
CI is only red because github is stupid and tries to merge before running CI, instead of testing the latest commit. This can be fixed in the github workflow file, maybe I should open a PR for that later.

Will merge later (need to handle the `HintD` declaration).

**Comment by rossberg** [comment]
Should we just annotate every node in the IL AST with its type?

**Comment by rossberg** [comment]
Yes, that's what I was thinking. And then we can remove all type annotations inside exp constructors. Not sure about making it optional, is there indication that it might be difficult in places?

**Comment by nomeata** [comment]
Not yet, we can try without option and I'll complain if it's hard somewhere.

**Comment by nomeata** [comment]
Or we could do what GHC does: only annotate enough leafs (variables and functions, I assume) so that one can infer the type for all expressions without context. A bit less bookkeeping, and performance is not super critical here.

**Comment by rossberg** [comment]
Okay, I implemented this. Also removed the now redundant annotations on DotE, CaseE, and DotP again.

It was slightly painful to construct the types in the sideconditions phase, because that required a variable environment. We can make them optional if that turns out to be too dreadful in the future.

**Comment by nomeata** [comment]
Cool, will update this pass accordingly in a few days (unless someone beats me to it)

**Comment by nomeata** [comment]
Updated to use the type information in the AST annotations.

**Review by rossberg — APPROVED** [review]

File: spectec/src/middlend/unthe.ml [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  let eqns2, x2' = t2 n x2 in
+  let eqns3, x3' = t3 n x3 in
+  eqns1 @ eqns2 @ eqns3, k (x1', x2', x3')
+
+(* Expr traversal *)
+
+let rec t_exp n e : eqns * exp =
+  (* Descend first using t_exp2, and then see if we have to pull out the current expression *)
+  let eqns, e' = t_exp2 n e in
+  match e.it with
```
Comment (matijapretnar) [inline spectec/src/middlend/unthe.ml:90]: @nomeata - Was this supposed to be `e'.it` instead of `e.it`?

Comment (nomeata) [reply spectec/src/middlend/unthe.ml:90]: Oh, quite likely. We probably don’t have nested `the` so far. Thanks for spotting!

**Pull request merged**